blocks/settlement.go

```diff
@@ -0,0 +1,165 @@
+// during block building and defines `x` as the block height of
+// `b.LastSettled()` while `y` as the height of the argument passed to this
+// method.
+//
+// The argument is typically the return value of [LastToSettleAt], where that
+// function receives b as the parent. See the Example.
+func (b *Block) IfChildSettles(lastSettledOfChild *Block) []*Block {
+	return settling(b.LastSettled(), lastSettledOfChild)
+}
+
+// settling returns all the blocks after `lastOfParent` up to and including
+// `lastOfCurr`, each of which are expected to be the block last-settled by a
+// respective block-and-parent pair. It returns an empty slice if the two
+// arguments have the same block hash.
+func settling(lastOfParent, lastOfCurr *Block) []*Block {
+	var settling []*Block
+	// TODO(arr4n) abstract this to combine functionality with iterators
+	// introduced by @StephenButtolph.
+	for s := lastOfCurr; s.ParentBlock() != nil && s.Hash() != lastOfParent.Hash(); s = s.ParentBlock() {
+		settling = append(settling, s)
+	}
+	slices.Reverse(settling)
+	return settling
+}
+
+// LastToSettleAt returns (a) the last block to be settled at time `settleAt` if
+// building on the specified parent block, and (b) a boolean to indicate if
+// settlement is currently possible. If the returned boolean is false, the
+// execution stream is lagging and LastToSettleAt can be called again after some
+// indeterminate delay.
+//
+// See the Example for [Block.IfChildSettles] for one usage of the returned
+// block.
+func LastToSettleAt(settleAt uint64, parent *Block) (*Block, bool) {
+	// These variables are only abstracted for clarity; they are not needed
+	// beyond the scope of the `for` loop.
+	var block, child *Block
+	block = parent // therefore `child` remains nil because it's what we're building
+
+	// The only way [Block.ParentBlock] can be nil is if `block` was already
+	// settled (see invariant in [Block]). If a block was already settled then
+	// only that or a later (i.e. unsettled) block can be returned by this loop,
+	// therefore we have a guarantee that the loop update will never result in
+	// `block==nil`.
+	for ; ; block, child = block.ParentBlock(), block {
+		if startsNoEarlierThan := block.Time(); startsNoEarlierThan > settleAt {
+			continue
+		}
+
+		if t := block.executionExceededSecond.Load(); t != nil && *t >= settleAt {
+			continue
+		}
+		if e := block.execution.Load(); e != nil {
+			if e.byGas.CompareUnix(settleAt) > 0 {
+				// There may have been a race between this check and the
+				// execution-exceeded one above, so we have to check again.
+				continue
+			}
+			return block, true
+		}
+
+		// TODO(arr4n) more fine-grained checks are possible for scenarios where
+		// (a) `block` could never execute before `settleAt` so we would
+		// `continue`; and (b) `block` will definitely execute in time and
+		// `child` could never, in which case return `nil, false`.
+		_ = child
+
+		return nil, false
```

Comment:
While this function does seem correct, I feel like we will probably need to address this TODO before putting this into production.

I think we should be able to perfectly implement this function without too much added complexity, right?

Reply:
Merged #19 to address this. Thank you.

> I think we should be able to perfectly implement this function without too much added complexity, right?

I'm not sure about _perfectly_ but getting much closer now. I added a new TODO to explain how to eke out the last bits of efficiency.